issue: 2670645 Add xlio extra api support

igor-ivanov opened the pull request.

igor-ivanov committed:
issue: 2670645 Add xlio extra api support

Signed-off-by: Igor Ivanov <igori@nvidia.com>

igor-ivanov committed:
issue: 2670645 Support XLIO in configure

Signed-off-by: Igor Ivanov <igori@nvidia.com>

igor-ivanov committed:
issue: 2670645 Add --load-xlio, --xliorxfiltercb, --xliozcopyread options

Signed-off-by: Igor Ivanov <igori@nvidia.com>

AlexanderGrissik commented on src/vma-redirect.cpp:
```diff
@@ -79,12 +79,12 @@ daemon_fptr_t fn_daemon = NULL;
 
 ////////////////////////////////////////////////////////////////////////////////
-#define VMA_LOG_CB_ENV_VAR "VMA_LOG_CB_FUNC_PTR"
+#define XLIO_LOG_CB_ENV_VAR "VMA_LOG_CB_FUNC_PTR"
```
Maybe change the  name itself to "XLIO_LOG_CB_FUNC_PTR"?

AlexanderGrissik commented on src/sockperf.cpp:
```diff
@@ -262,10 +262,10 @@ static const AOPT_DESC common_opt_desc[] = {
+      "Load XLIO dynamically even when LD_PRELOAD was not used." },
     { OPT_RATE_LIMIT, AOPT_ARG, aopt_set_literal(0), aopt_set_string("rate-limit"),
-      "use rate limit (packet-pacing), with VMA must be run with VMA_RING_ALLOCATION_LOGIC_TX "
+      "use rate limit (packet-pacing), with XLIO must be run with VMA_RING_ALLOCATION_LOGIC_TX "
```
VMA_RING_ALLOCATION_LOGIC_TX to XLIO_RING_ALLOCATION_LOGIC_TX ?

AlexanderGrissik commented on src/sockperf.cpp:
```diff
@@ -2073,7 +2073,7 @@ static char *display_opt(int id, char *buf, size_t buf_size) {
 
 //#define ST_TEST
-// use it with command such as: VMA ./sockperf -s -f conf/conf.inp --vmarxfiltercb
+// use it with command such as: VMA ./sockperf -s -f conf/conf.inp --xliorxfiltercb
```
"... as: VMA ..." to "... as : XLIO ..." ?

AlexanderGrissik commented on src/sockperf.cpp:
```diff
@@ -3396,8 +3396,8 @@ packet pace limit = %d",
+            s_user_params.is_xliozcopyread, s_user_params.mc_loop_disable, s_user_params.mc_ttl,
             s_user_params.uc_reuseaddr, s_user_params.tcp_nodelay,
             s_user_params.client_work_with_srv_num, s_user_params.b_server_reply_via_uc,
             s_user_params.b_server_dont_reply, s_user_params.b_server_detect_gaps,
```
See lines 3422, 3418
Also didn't find VMA_LIBRARY_MAJOR definition, in libxlio there something like PRJ_LIBRARY_MAJOR.

AlexanderGrissik commented on src/vma-redirect.cpp:
```diff
@@ -95,11 +95,11 @@ static vma_log_cb_t vma_log_get_cb_func() {
+        unsetenv(XLIO_LOG_CB_ENV_VAR);
         return false;
     }
 
```
see line 176, not sure if it should stay like this ...

AlexanderGrissik reviewed (changes requested):
Reviewed all the files and added some comments.
Also there is some modification needed in doc/main.dox which still describes old flags. like --load-vma.

igor-ivanov committed:
issue: 2670645 Fix printf() NULL argument issue

Signed-off-by: Igor Ivanov <igori@nvidia.com>

igor-ivanov committed:
issue: 2618091 Option --tls can not be used with --load-xlio

It is sockperf limitation.
Reason:
--tls option force using TLS API that can call socket API.
But sockperf just can redirect sockper internal sockect function calls
to libxlio with --load-xlio option.
LD_PRELOAD usage overcome this.

Signed-off-by: Igor Ivanov <igori@nvidia.com>

swx-jenkins3 commented:
Can one of the admins verify this patch?

igor-ivanov commented:
Closed because final solution as https://github.com/Mellanox/sockperf/pull/198 is merged
